docs/developer-docs/security/security-best-practices/data-integrity-and-authenticity.mdx

```diff
@@ -2,25 +2,628 @@
+
+### Using certified variables for secure queries 
+In certain use cases, there is a third option whereby query results can return data that has been certified by the subnet in an earlier update call. This is the concept of certified data and it requires changes to the update call to create the certification, query call to return the certificate and front-end to verify the certificate. Using certified data provides the best of both worlds with query-like response times and update-like certified responses. This forms the core of [certified variables](https://internetcomputer.org/docs/current/tutorials/developer-journey/level-2/2.2-advanced-canister-calls/#certified-variables). 
+
+Some examples of certified variables are asset certification in [Internet Identity](https://github.com/dfinity/internet-identity/blob/b29a6f68bbe5a49d048e12bc7a3263a9f43d080b/src/internet_identity/src/main.rs#L775-L808), [NNS dapp](https://github.com/dfinity/nns-dapp/blob/372c3562127d70c2fde059bc9c268e8ae858583e/rs/src/assets.rs#L121-L145), or the [canister signature implementation in Internet Identity](https://github.com/dfinity/internet-identity/blob/main/src/internet_identity/src/signature_map.rs).
+
+:::tip
+Certified variables is an advanced feature which require careful implementation of authenticated data structures and verification on the canister and client side respectively. **If the client doesn't require fast response times, call the query method as an update call (replicated query).** The response would be certified by the subnet and a single malicious or boundary node can't modify the response.
+:::
+
+:::tip
+ICP also provides replica signed queries, where query responses are signed by the answering replica node, however it doesn't have the same security guarantees an `update` call and only protects from malicious boundary nodes. Replica signed queries are enabled by default on both `agent-rs` and `agent-js`. [Read more about the feature](https://internetcomputer.org/blog/features/replica-signed-queries).
+:::
+
+### What is certified data?
+A canister can store a small amount of data (32 bytes) during an update call in the state tree under `certified_data` which will be certified by the subnet. During a query call, the canister can obtain a certificate about that data, which can be verified against the IC root public key. Combining these features, a canister should be able to certify a variable in a query response with this [design](https://medium.com/dfinity/how-internet-computer-responses-are-certified-as-authentic-2ff1bb1ea659). 
```

Comment:
I've written an alternative introduction. It also explains the issue with calculations in query calls. WDYT?

> For a subnet to certify data, it has to create a signature over all certified data during an update call. Each update call that modifies the data needs to recreate this signature. However, a canister won't simply sign an arbitrarily large amount of data in complex data structure. It is limited to signing just 32 bytes which will end up in the subnet state tree under `certified_data`. Since this is rarely enough, it is up to the developer to reduce the data into this 32 bytes. Naturally, this can be done by computing a hash of the data structure. This is also why certified data is difficult to implement. Depending on your data structures, you will need to develop a different kind of hashing function.
> 
> Subsequent query calls can return the data as-is, including the signature which the front-end can verify. This means that data aggregation or other calculations can't be done in query calls as there would be no way to produce a signature over that newly created data. There are two workarounds, either this data is precomputed in the update call or all raw data is sent to the front-end which verifies it and does the calculations.

Reply:
I will try to adapt this.